graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,184 @@
+            )
+        )
+
+
+@dataclass(frozen=True)
+class VertexPartitionPlan:
+    """Plan to split the query at a certain vertex into a certain number of pages."""
+
+    # field names leading to the vertex to be split
+    query_path: Tuple[str, ...]
+
+    # field to use for pagination
+    pagination_field: str
+
+    # The number of subdivisions intended for this vertex
+    number_of_splits: int
+
 
+@dataclass(frozen=True)
+class PaginationPlan:
+    """Plan to split the query with a combination of VertexPartitionPlans."""
 
-# The intent to split the query at a certain vertex into a certain number of pages.
-VertexPartition = namedtuple(
-    "VertexPartition",
-    (
-        "query_path",  # Tuple[field name : str] leading to the vertex to be split
-        "number_of_splits",  # The number of subdivisions intended for this vertex
-    ),
-)
+    vertex_partitions: Tuple[VertexPartitionPlan, ...]
 
 
-# The intent to split the query with a combination of VertexPartitions
-PaginationPlan = namedtuple("PaginationPlan", ("vertex_partitions",))  # List[VertexPartition]
+def get_plan_page_count(plan: PaginationPlan) -> int:
+    """Return the number of pages that a PaginationPlan would generate."""
+    number_of_pages = 0
+    for vertex_partition in plan.vertex_partitions:
+        number_of_pages *= vertex_partition.number_of_splits
+    return number_of_pages
 
 
-# TODO(bojanserafimov): Make this function return a best effort pagination plan
-#                       when a good one is not found instead of returning None.
-def try_get_pagination_plan(schema_info, query_ast, number_of_pages):
-    """Make a PaginationPlan for the given query and number of desired pages if possible."""
+def get_best_vertex_partition_plan(
+    schema_info: QueryPlanningSchemaInfo, pagination_node: str, number_of_pages: int
+) -> Tuple[Optional[VertexPartitionPlan], Tuple[PaginationAdvisory, ...]]:
+    """Get the best available VertexPartitionPlan on this node or None if no plan exists."""
+    pagination_field = schema_info.pagination_keys.get(pagination_node)
+    if pagination_field is None:
+        return None, (PaginationFieldNotSpecified(pagination_node),)
+
+    # Specifying an unsupported pagination_field is a compiler bug
+    if not field_supports_range_reasoning(schema_info, pagination_node, pagination_field):
+        vertex_type = schema_info.schema.get_type(pagination_node)
+        field_type_name = vertex_type.fields[pagination_field].type.name
+        raise AssertionError(
+            u"Cannot paginate on {}.{} because pagination on {} is not supported ".format(
+                pagination_node, pagination_field, field_type_name
+            )
+        )
+
+    # Ideally, we paginate into the desired number of pages, and don't advise for any changes.
+    # These two values are mutated in the code below if this vertex is not ideal in any way.
```

Comment:
Nit: the values themselves aren't mutated (they are immutable), instead the variables reference new values.
```suggestion
    # These two variables are updated in the code below if this vertex is not ideal in any way.
```